bls377/g1.go

```diff
@@ -469,6 +486,113 @@ func (p *G1Jac) Double() *G1Jac {
 	return p
 }
 
+// queryNthBit returns the i-th bit of s
+func queryNthBit(s fr.Element, i int) uint64 {
+	limb := i / 64
+	offset := i % 64
+	b := (s[limb] >> offset) & 1
+	return b
+}
+
+// doubleandadd algo for exponentiation
+// n=number of bits of the scalar s
+func (p *G1Jac) doubleandadd(curve *Curve, a *G1Affine, s fr.Element, n int) *G1Jac {
+
+	var res G1Jac
+	res.Set(&curve.g1Infinity)
+
+	for i := n - 1; i >= 0; i-- {
+		b := queryNthBit(s, i)
+		res.Double()
+		if b == 1 {
+			res.AddMixed(a)
+		}
+	}
+	p.Set(&res)
+	return &res
+}
+
+// ScalarMulEndo performs scalar multiplication
+// using the endo phi(p=(x,y))=(ux,y) where u is a 3rd root of 1,
+// phi(P) = lambda*P
+// u = 80949648264912719408558363140637477264845294720710499478137287262712535938301461879813459410945
+// lambda = 91893752504881257701523279626832445440 (129 bits)
+// s1, s2 are scalars such that s1*lambda+s2 = s
+// s1 on 129 bits
+// s2 on 127 bits
+func (p *G1Jac) ScalarMulEndo(curve *Curve, a *G1Affine, s fr.Element) G1Jac {
+
+	// operation using big int
+	var lambda, _s, _s1, _s2 big.Int
+	lambda.SetString("91893752504881257701523279626832445440", 10)
```

Comment:
Why are we doing string-to-integer conversion of a constant in every single call to `ScalarMulEndo`?  (Same for line 536.)  We should hard-code this constant instead.  Even if we are stuck using `big.Int` there _must_ be a faster way to accomplish this.

Is `big.Int` really the best choice here?  Seems to me that `fr` is preferable.  Regarding your [earlier comment](https://github.com/ConsenSys/gurvy/pull/3#issuecomment-621368191): seems worth it to add Euclidean division to `goff` for use here.  Perhaps we should open a new `goff` issue for this task.

Reply:
It's not so bad performance wise. The algo is still faster than the normal scalar double and add. I'll for Gautam to implement euclidean division in goff, but I am not sure it's needed (at least it's not urgent for this algo).